### PR TITLE
fix(codex): improve transcript parsing, tool argument formatting, and compaction support

### PR DESCRIPTION
**Pull request by rangemer333-cell**
## Summary
- Add web_search duration fix using lastNonWebSearchTs approximation (web_search_call only has end time, approximate start from preceding events)
- Support custom_tool_call, tool_search_call, and message response_item parsing in transcript-parser
- Refactor appendMissingTranscriptToolEvents from Set to Map for proper merging in hook-processor
- Add formatCodexToolArguments for Bash workdir normalization and apply_patch handling
- Change UserPromptSubmit event.name from `llm.request` to `other` (not an actual LLM request)
- Support llm_full_input_messages for compaction scenarios in react-step-builder
- Merge duplicate turn_context events
- Add 20 passing tests covering all new functionality

## Test plan
- [x] `npx vitest run tests/unit/hooks/codex/` — 20/20 tests passed
- [ ] Manual verification with real Codex sessions containing web_search, custom_tool_call, and compaction events

### CODE REVIEW

File: assets/hooks/codex-hook-processor.mjs [hunk 1]
```diff
@@ -107,20 +107,78 @@ function requireSessionId(event, stage = 'cmd') {
...
+
+  if (normalizedName === 'apply_patch') {
+    if (hookArgs.command !== undefined) return hookArgs;
+    if (transcriptArgs.command !== undefined) return transcriptArgs;
+    if (typeof hookInput === 'string') return { command: hookInput };
+    if (typeof transcriptInput === 'string') return { command: transcriptInput };
+    if (isPlainObject(hookInput)) return hookInput;
+    if (isPlainObject(transcriptInput)) return transcriptInput;
+    const command = hookInput ?? transcriptInput;
+    return command !== undefined ? { command } : command;
```
Comment (linrunqi08) [inline assets/hooks/codex-hook-processor.mjs]: **Bug (edge case):** When both `hookInput` and `transcriptInput` are `null`, this line returns `{ command: null }` instead of the likely-intended `undefined` or `null`.

Trace: `hookInput ?? transcriptInput` → `null ?? null` → `null`. Then `null !== undefined` → `true`, so it returns `{ command: null }`.

While rare for `apply_patch` (which normally receives string input), if a transcript entry has no `input` field, `parseMaybeJsonValue(payload.input)` returns `undefined` which then becomes `null` via `??` chains. Downstream `toJsonValue({ command: null })` would emit a malformed tool argument.

Suggested fix:
```js
return command != null ? { command } : hookInput;
```

File: assets/hooks/codex-hook-processor.mjs [hunk 2]
```diff
@@ -107,20 +107,78 @@ function requireSessionId(event, stage = 'cmd') {
...
+  }
   for (const event of transcriptToolEvents) {
     const key = `${event.type}:${event.tool_use_id || ''}`;
-    if (!seen.has(key)) {
+    const existing = seen.get(key);
+    if (existing) {
+      mergeTranscriptToolEvent(existing, event);
+    } else {
+      if (event.type === 'pre_tool_use') {
+        event.tool_input = formatCodexToolArguments(event.tool_name, event.tool_input, event.tool_name, event.tool_input);
```
Comment (linrunqi08) [inline assets/hooks/codex-hook-processor.mjs]: **Cleanup:** This calls `formatCodexToolArguments(event.tool_name, event.tool_input, event.tool_name, event.tool_input)` — passing identical values as both hook-source and transcript-source params. The function's 4-arg design is meant to merge data from two different sources (hook vs transcript), but here both pairs are the same event.

The normalization still works (extracts `command`/`workdir` for Bash, wraps string input for `apply_patch`), but the duplicate-param call is confusing and suggests the merge intent isn't being realized here.

Consider either:
- Extracting a simpler `normalizeToolInput(toolName, input)` for the single-source case
- Or just documenting that the 4-arg call is intentional for normalization-only (no merge)

File: assets/hooks/codex/transcript-parser.mjs [hunk 3]
```diff
@@ -321,6 +380,62 @@ export function parseTranscript(transcriptPath, byteOffset = 0, initialLastUsage
           // 子 agent 的工具调用在子 transcript 中，不会出现在这里。
           parentToolCallIds.add(callId);
         }
+      } else if (itemType === 'custom_tool_call') {
```
Comment (linrunqi08) [inline assets/hooks/codex/transcript-parser.mjs:383]: **Altitude/maintainability:** The `custom_tool_call`, `web_search_call`, and `tool_search_call` handlers share a nearly identical pattern: extract callId → create `pre_tool_use` event → add to `pendingToolCalls`/`toolEvents`/`parentToolCallIds` → (for output) create `post_tool_use` event → delete from `pendingToolCalls`.

The only variation per type is which payload fields map to `callId`, `name`, `input`, and `output`. A table-driven approach would reduce each new type from ~15 lines to a config entry:

```js
const RESPONSE_ITEM_TOOL_MAP = {
  function_call:     { callId: ['call_id','id'], name: 'name', input: 'arguments', parse: true },
  custom_tool_call:  { callId: ['call_id','id'], name: 'name', input: 'input',     parse: false },
  tool_search_call:  { callId: ['call_id','id'], name: null,   input: 'arguments', parse: false, defaultName: 'tool_search' },
  // ...
};
```

Not blocking, but when Codex adds more types (`code_interpreter_call`, `file_search_call`, etc.), the else-if chain will grow linearly.

File: assets/hooks/codex-hook-processor.mjs [hunk 4]
```diff
@@ -107,20 +107,78 @@ function requireSessionId(event, stage = 'cmd') {
...
+    if (isPlainObject(transcriptInput)) return transcriptInput;
+    const command = hookInput ?? transcriptInput;
+    return command !== undefined ? { command } : command;
+  }
+
+  if (isPlainObject(transcriptInput)) return transcriptInput;
+  return hookInput ?? transcriptInput;
+}
+
+function isPlainObject(value) {
```
Comment (linrunqi08) [inline assets/hooks/codex-hook-processor.mjs]: **Reuse:** This `isPlainObject` check (`value !== null && typeof value === 'object' && !Array.isArray(value)`) duplicates the existing `asRecord(value)` in `agent-event-normalizer.mjs`. The same inline pattern appears in at least 5 other locations across the hooks directory (`tool-utils.mjs`, `hook-processor-base.mjs`, `cursor-hook-processor.mjs`).

Consider importing `asRecord` from `agent-event-normalizer.mjs` and adapting for the boolean return type, or extracting a shared `isPlainObject` to consolidate the pattern.

**Review by linrunqi08 — COMMENTED** [review]
## Code Review — PR #13

Reviewed the full diff across all 5 files. Overall the PR is well-structured with solid test coverage (20 tests). Found 4 findings worth addressing (1 edge-case bug, 3 cleanup/maintainability items). No critical blocking issues.

File: assets/hooks/codex-hook-processor.mjs (hunk 1 above)
Comment (rangemer333-cell) [reply assets/hooks/codex-hook-processor.mjs]: 已修复。现在 apply_patch 在 hook/transcript 输入都为空时不会再包装成 { command: null }，避免下游生成畸形 tool arguments。对应逻辑已调整为仅在 command != null 时包装，并补充了单测覆盖该边界场景。

File: assets/hooks/codex-hook-processor.mjs (hunk 2 above)
Comment (rangemer333-cell) [reply assets/hooks/codex-hook-processor.mjs]: 已调整。这里原本确实混淆了“单源归一化”和“hook/transcript 双源合并”两个语义。现在拆成了 normalizeCodexToolArguments(...) 和 mergeCodexToolArguments(...)，transcript-only 场景走单源归一化，hook + transcript 场景走双源合并，可读性更清楚。

File: assets/hooks/codex/transcript-parser.mjs (hunk 3 above)
Comment (rangemer333-cell) [reply assets/hooks/codex/transcript-parser.mjs:383]: 认同这个方向。目前这次改动先保持现有结构，避免在本轮 reviewer cleanup 中扩大重构范围。后续如果继续补 code_interpreter_call、file_search_call 等类型，可以把这些 response item handler 收敛成 table-driven 配置，减少分支链增长。

File: assets/hooks/codex-hook-processor.mjs (hunk 4 above)
Comment (rangemer333-cell) [reply assets/hooks/codex-hook-processor.mjs]: 认同可以进一步收敛。不过这次新增的判断被限制在 tool-arguments.mjs 内部，先避免跨模块调整公共 helper 带来额外影响。后续可以单独做一轮 hooks 目录内的 plain-object 判断清理，把重复逻辑统一到共享工具函数里。

**Review by linrunqi08 — APPROVED** [review]
